Added new Authorization Roles (#360)
epinio/docs: docs/references/authorization.md

# Authorization

Since version **1.11.0** Epinio is shipped with a new authorization layer with two default roles: **admin** and **user**.
Roles can be "namescoped", and they can be also customized with different permissions.

> A user with the admin role will have the permission to perform any operation. A standard user will have only read permissions, but it also has te permissions to create namespaces. When a user creates a namespace, it will have automatically the admin permission for it.

Reviewer: ```suggestion
A user with the admin role will have the permission to perform any operation. A standard user will have only read permissions, but also has the permissions to create namespaces. When a user creates a namespace, they will automatically have the admin permission for it.
```

EOF
```

## Roles

An Epinio role is a Kubernetes ConfigMap, with the `epinio.io/role` reserved label.

The following yaml shows you the default `user` role:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    epinio.io/role: "true"
  name: "epinio-user-role"
  namespace: {{ $.Release.Namespace }}
data:
  id: user
  name: "Epinio User Role"
  default: "true"
  actions: |
    namespace
    app_read
    configuration_read
    service_read
    gitconfig_read
```

### Fields

| Key     | Description 
|---------|-------------
| id      | The ID of the Role
| name    | A friendly name for the Role
| default | (optional) if set to _true_ the role will be the one selected as default if no other roles were assigned to the user
| actions | The actions the roles can perform

## Actions

Each Role can perform some actions defined in the __actions__ field (newline separated). These actions are hardcoded in Epinio, and some of them depends on other actions. Enabling an action with a dependency will enable also the dependency automatically.

### Namespace

> These actions will enable operations on the Namespace commands and resources.

Reviewer 2: ```suggestion
These actions enable operations on Namespace commands and resources.
```

| Action ID         | Description 
|-------------------|-------------
| `namespace_read`    | Read permissions (list, show)
| `namespace_write`   | Write permissions (create, delete)<br/>Depends on: `namespace_read`
| `namespace`         | All the above<br/>Depends on: `namespace_read`, `namespace_write`

### App

> These actions will enable operations on the App commands and resources. They will also enable commands related to the AppChart (`epinio app chart`) and the application environment variables.

Reviewer 2: ```suggestion
These actions enable operations on App commands and resources. They also enable commands related to  AppCharts (`epinio app chart`) and application environment variables.
```

| Action ID       | Description 
|-----------------|-------------
| `app_read`        | Read permissions (app list and show, env list and show)
| `app_logs`        | Read application logs
| `app_write`       | Write permissions (app create, delete, push, export, stage, env set and unset)<br/>Depends on: `app_read`, `app_logs`
| `app_exec`        | Perform an exec into a running application
| `app_portforward` | Open a tunnel with the `port-forward` command
| `app`             | All the above<br/>Depends on: `app_read`, `app_logs`, `app_write`, `app_exec`, `app_portforward`

### Configuration

> These actions will enable operations on the Configuration commands and resources. Be aware that to bind a configuration you will still need the `app_write` permission.

Reviewer 2: ```suggestion
These actions enable operations on Configuration commands and resources. Be aware that to bind a configuration you still need the `app_write` permission as well.
```

| Action ID           | Description 
|----------------------|-------------
| `configuration_read`  | Read permissions (list, show)
| `configuration_write` | Write permissions (create, delete)<br/>Depends on: `configuration_read`
| `configuration`       | All the above<br/>Depends on: `configuration_read`, `configuration_write`

### Service

> These actions will enable operations on the Service commands and resources. 

Reviewer 2: ```suggestion
These actions enable operations on Service commands and resources. 
```

| Action ID             | Description 
|-----------------------|-------------
| `service_read`        | Read permissions (list, show)
| `service_write`       | Write permissions (create, delete, bind, unbind)<br/>Depends on: `service_read`
| `service_portforward` | Open a tunnel with the `port-forward` command
| `service`             | All the above<br/>Depends on: `service_read`, `service_write`, `service_portforward`

### Gitconfig

> These actions will enable operations on the Gitconfig commands and resources.

Reviewer 2: ```suggestion
These actions enable operations on Gitconfig commands and resources.
```

| Action ID         | Description 
|-------------------|-------------
| `gitconfig_read`    | Read permissions (list, show)
| `gitconfig_write`   | Write permissions (create, delete)<br/>Depends on: `gitconfig_read`
| `gitconfig`         | All the above<br/>Depends on: `gitconfig_read`, `gitconfig_write`

### Export Registries

> This action will enable operations on the Export Registries commands and resources. Only read operations are available.

Reviewer 2: ```suggestion
This action enable operations on Export Registries commands and resources. Only read operations are available.
```
